### ISSUE TITLE
ttnn.rsqrt_bw returns +inf at input == 0 where torch returns -inf (sign error)

### ISSUE THREAD

**Issue by Lanochu** [tenstorrent/tt-metal#56380]
## Summary

At `input == 0`, `rsqrt_bw` overrides its computed result with `+inf` regardless of the sign of the
incoming gradient. The derivative of `rsqrt` is **negative** (`d/dx x^(-1/2) = -0.5 x^(-3/2)`), so
for a positive incoming gradient the correct limit is `-inf`. The op returns the wrong sign for
every positive gradient at a zero input.

| input | grad | torch (golden) | ttnn.rsqrt_bw | |
|---:|---:|---:|---:|---|
| 0.0 | 1.0 | **-inf** | **+inf** | wrong sign |
| 0.0 | -1.0 | +inf | +inf | ok |
| 0.0 | 0.0 | nan | nan | ok |
| 4.0 | 1.0 | -0.0625 | -0.0625 | ok |
| 0.25 | 2.0 | -8.0000 | -8.0000 | ok |
| -1.0 | 1.0 | nan | nan | ok |

For comparison, `sqrt_bw` in the same file handles the identical situation correctly, because it
branches on the gradient's sign:

```cpp
where(eqz(input) && ltz(grad), -inf, ...);
where(eqz(input) && gtz(grad), +inf, ...);
```

A wrong-signed infinity is worse than a NaN in training: it does not announce itself, and after a
subsequent multiply or clamp it can become a large finite value pushing the update in exactly the
wrong direction.

## Root cause

`ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L482-L514`:

```cpp
ttnn::rsqrt(input, false, output_mem_config, input_grad);
ttnn::power(input_grad.value(), 3, output_mem_config, input_grad);
ttnn::multiply(ttnn::multiply(grad, input_grad.value(), ...), -0.5f, ..., input_grad);

where(ttnn::eqz(input, ...), t_inf, input_grad.value(), ..., input_grad);   // <-- unconditional +inf
where(ttnn::ltz(input, ...), t_nan, input_grad.value(), ..., input_grad);
where(ttnn::logical_and(ttnn::eqz(input, ...), ttnn::eqz(grad, ...)), t_nan, ..., input_grad);
```

The arithmetic on the first three lines already carries the sign correctly: the `-0.5f` factor is
what makes the result negative for positive `grad`. The fourth line then discards that and
substitutes a bare `+inf`. The subsequent `eqz(input) && eqz(grad) -> NaN` line shows the author was
thinking about the `0 * inf` case, but the sign of the non-zero-gradient case was not considered.

## Why tests miss it

- `tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_rsqrt.py` builds the
  input with `data_gen_with_range(shapes, -201, 199)`, i.e.
  `torch.rand(shape).bfloat16() * 400 - 201`. Hitting `input == 0` requires
  `rand.bfloat16() == 201/400 = 0.5025`, which is **not a representable bfloat16 value** — the
  spacing near 0.5 is `2^-9 = 0.001953125`, and the neighbouring values map to `-0.21875` and
  `+0.5625`. Across all three parametrised shapes the test produces **zero** inputs equal to 0,
  so the branch is never executed.
- `tests/sweep_framework/sweeps/eltwise/unary_backward/rsqrt_bw/rsqrt_bw.py` draws the input from
  `torch_random(low=0, high=100)` cast to bfloat16, which likewise essentially never yields exactly
  `0`, and asserts `check_with_pcc(..., 0.99)`.
- Even if the branch were reached, `get_atol_rtol_pcc`
  (`ttnn/tt_lib/_internal/comparison_funcs.py#L41-L55`) replaces every NaN and **both signed
  infinities** with `0` before correlating:

  ```python
  # For now, mask all infs and nans so that we check the rest... TODO
  golden[isnan | isinf | isneginf] = 0
  calculated[isnan | isinf | isneginf] = 0
  ```

  so the golden's `-inf` and the device's `+inf` both become `0` and compare equal. A sign error at
  an infinity is structurally invisible to every PCC-based test in this suite.

## Proposed fix

Make the substitution sign-aware, using the idiom already present in `log_bw` in the same file
(`unary_backward.cpp#L840-L858`), which multiplies an infinity by `ttnn::sign(grad)`:

```cpp
// d/dx rsqrt(x) -> -inf as x -> 0+, so the result carries the OPPOSITE sign to grad.
where(
    ttnn::eqz(input, ...),
    ttnn::multiply(ttnn::sign(grad, ...), -std::numeric_limits<float>::infinity(), std::nullopt, ...),
    input_grad.value(), ..., input_grad);
```

`grad == 0` then yields `0 * -inf`; keep the existing `eqz(input) && eqz(grad) -> NaN` line so that
case stays explicit and does not depend on `ttnn::sign(0)`.

I did not propose simply deleting the override — although the preceding arithmetic looks like it
would produce `-inf` naturally, that depends on what `ttnn::rsqrt(0)` and `ttnn::power(inf, 3)`
return on device, which I could not confirm. The explicit form above does not depend on either.

For tests, `input == 0` should be pinned as an explicit value rather than sampled, and asserted with
an exact/sign-aware comparison; PCC cannot distinguish `+inf` from `-inf`.

## Validation status

The device column is a transcription of the composite op's `where` chain, not a silicon run; the
torch column is real `torch.rsqrt` autograd output, which is the registered golden. The defect is
visible directly in the source. I have not run this on a Wormhole or Blackhole part and would want
on-device confirmation as the first step of the fix.

<details>
<summary><b>Standalone reproduction</b> (torch only, no Tenstorrent hardware required)</summary>

```python
"""
Reproduction for ttnn.rsqrt_bw returning +inf where torch returns -inf at input == 0.

The 'ttnn' column transcribes the `where` chain in
ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp::rsqrt_bw.
The 'torch' column is real autograd output -- the registered golden.

Requires: torch.   Run:  python rsqrt_bw_repro.py
"""
import torch
inf = float("inf"); nan = float("nan")

def ttnn_rsqrt_bw(x, g):
    r = -0.5 * g * (x ** -1.5) if x > 0 else nan
    if x == 0:              r = inf     # where(eqz(input), +inf, ...)   <-- ignores sign of grad
    if x < 0:               r = nan     # where(ltz(input), NaN, ...)
    if x == 0 and g == 0:   r = nan     # where(eqz(input) & eqz(grad), NaN, ...)
    return r

def ttnn_sqrt_bw(x, g):                 # same file, handled correctly -- shown for contrast
    import math
    r = g / (2 * math.sqrt(x)) if x > 0 else nan
    if x <= 0:              r = nan
    if x == 0 and g < 0:    r = -inf
    if x == 0 and g > 0:    r = inf
    return r

def torch_bw(fn, x0, g0):
    t = torch.tensor([x0], dtype=torch.float64, requires_grad=True)
    fn(t).backward(torch.tensor([g0], dtype=torch.float64))
    return t.grad.item()

def same(a, b): return (a != a and b != b) or a == b

for name, tfn, kfn in [("rsqrt_bw", torch.rsqrt, ttnn_rsqrt_bw), ("sqrt_bw (correct, for contrast)", torch.sqrt, ttnn_sqrt_bw)]:
    print(f"\n=== {name} ===")
    print(f"   {'input':>7} {'grad':>6} {'torch':>12} {'ttnn':>12}   verdict")
    for x0, g0 in [(0.0,1.0),(0.0,-1.0),(0.0,0.0),(4.0,1.0),(0.25,2.0),(-1.0,1.0)]:
        ref, got = torch_bw(tfn, x0, g0), kfn(x0, g0)
        print(f"   {x0:7.2f} {g0:6.1f} {ref:12.4f} {got:12.4f}   {'ok' if same(ref,got) else 'MISMATCH'}")

print("\n\n=== why the unit test never reaches input == 0 ===")
print("test_backward_rsqrt.py: data_gen_with_range(shape, -201, 199)")
print("  = torch.rand(shape).bfloat16() * 400 - 201")
print("  input == 0 requires rand.bfloat16() == 201/400 = 0.5025\n")
b = torch.tensor([0.5025], dtype=torch.float32).bfloat16().float().item()
print(f"  nearest bfloat16 to 0.5025 is {b}  ->  {b}*400-201 = {b*400-201}")
torch.manual_seed(0)
for shape in [(1,1,32,32),(1,1,320,384),(1,3,320,384)]:
    inp = torch.rand(shape).bfloat16() * 400 - 201
    print(f"  shape {str(shape):<16} elements={inp.numel():>8,}  inputs exactly 0: {int((inp==0).sum())}")
print("\n  and get_atol_rtol_pcc maps BOTH +inf and -inf to 0 before correlating,")
print("  so a sign error at an infinity cannot fail a PCC assertion.")

```

</details>

**Comment by Lanochu** [comment]
I have a fix for this ready and would like to be assigned before opening a PR, per the bounty
terms ("you must be assigned on GitHub to the issue for which you are submitting a pull request").

**The change** — make the substitution at `input == 0` sign-aware, using the idiom already present
in `log_bw` and `acosh_bw` in the same file:

```cpp
// d/dx rsqrt(x) = -0.5 * x^(-3/2) -> -inf as x -> 0+, so at input == 0 the result carries the
// sign OPPOSITE to grad. Substituting a bare +inf here dropped that sign, returning +inf where
// torch returns -inf for any positive incoming gradient.
where(
    ttnn::eqz(input, output_mem_config),
    ttnn::multiply(ttnn::sign(grad, output_mem_config), -t_inf, std::nullopt, output_mem_config),
    input_grad.value(),
    output_mem_config,
    input_grad);
```

The existing `ltz(input) -> NaN` and `eqz(input) && eqz(grad) -> NaN` lines are unchanged. Net diff
is 9 added / 1 removed lines in `unary_backward.cpp`.

**Regression test** added to `test_backward_rsqrt.py`, pinning `input == 0` against the registered
golden with an exact comparison. It deliberately does not use PCC: `get_atol_rtol_pcc` maps both
`+inf` and `-inf` to `0` before correlating, so a sign error at an infinity cannot fail a PCC
assertion. It also pins the value rather than sampling, because `data_gen_with_range(-201, 199)`
can never produce `input == 0` (that needs `rand.bfloat16() == 0.5025`, not a representable
bfloat16 value).

Verified against `torch.rsqrt` autograd on a grid covering `input == 0` with positive, negative and
zero gradients plus the smooth region: all cases match after the change. As noted in the issue, I
have not built tt-metal locally — I verified every API call against identical call sites in the
same translation unit, but your CI would be the first real compile.

Happy to open the PR as soon as it is assigned.

**Comment by Lanochu** [comment]
Closing as a duplicate — I missed the prior art when I filed this. Same defect as #53973 ("rsqrt_bw writes +inf at zero, inverting the sign for every positive gradient"), which reported it first.

Apologies for the extra triage load; I've cross-checked my remaining open reports against the existing corpus.
